### DISCUSSION TITLE
mutation typing

### DISCUSSION THREAD

**Discussion by hyusetiawan** [gqty-dev/gqty#1601] [Q&A]
im following the mutation for reach as shown here: https://gqty.dev/guides/react/write

but I am not getting the typing when I call the returned mutation function, like so:
<img width="729" alt="Screenshot 2023-08-26 at 1 59 37 PM" src="https://github.com/gqty-dev/gqty/assets/1062920/f285eca9-74b9-4e27-a149-08dce0fbbabe">
but inside the callback, I am able to see the typing properly:
<img width="557" alt="Screenshot 2023-08-26 at 1 59 09 PM" src="https://github.com/gqty-dev/gqty/assets/1062920/e387fbc1-fc5b-4bec-8d2c-2f2b8ac3e08e">

is there a way to type the `login` function?

**Comment by hyusetiawan** [comment]
I call `login` function like so:
```
 await login({
      args: {id}
    })
```
and the error message is : `Type '{ id: string | null; }' is not assignable to type 'never'.ts(2322)`

**Comment by vicary** [comment]
You'll need to explicitly type the callback from the 2nd parameter and on, those types will be inferred in the returning mutation function.

There are two common styles to type it:

## Direct imports

```ts
import { type Mutation, type Variables } from './gqty';

const [login, mutationStates] = useMutation((mutation, args: Variables<Mutation["login"]>) => {
  const token = mutation.login(args);
  return token;
});

// Later in your code
<Button onClick={() => {
  login({ username, password });
}}>Login</Button>
```

## Curated inputs

```ts
const [login, mutationStates] = useMutation((mutation, username: string, password: string) => {
  const token = mutation.login({ username, password });
  return token;
});

// Later in your code
<Button onClick={() => {
  login(username, password);
}}>Login</Button>
```

**Reply by hyusetiawan** [reply]
@vicary for both modes, where do you get the variable `mutation` in the button onclick?

the API is rather awkward, I am sure there is an engineering reasoning behind it, is it not possible to do:
```
const {login} = useMutation()
const [mutation, state] = login
```
so it is more consistent with the useQuery?

**Reply by vicary** [reply]
Sorry for the typo for both of the examples, there will be no `mutation` outside of the callback scope.

```diff
// Later in your code
<Button onClick={() => {
-  mutation.login(username, password);
+  login(username, password); 
}}>Login</Button>
```

I have updated the answer above to reflect that.

We have actually experimented with the `useQuery()` style for mutations, but users tend to keep the proxy object in states, e.g. `useState()`. Whenever a future render tries to access its properties, a new mutation will be fired. That's a pitfall too hostile for beginners, we eventually scrapped that.

**Reply by vicary** [reply]
On a second thought, it may work with modifications on the core logic. It's not trivial works though.

No guarantees, but I will on-and-off exploring the new idea in upcoming months.

**Reply by hyusetiawan** [reply]
I think it's feasible to preserve the current API signature and just create a wrapper around it to make it similar dev UX wise with `useQuery`, here, what do you think?

given the following generated mutation schema:
```
export interface Mutation {
  __typename?: 'Mutation'
  login: (args: { bla: Scalars['Int']; id: Scalars['String'] }) => Maybe<Session>
  logout?: Maybe<ScalarsEnums['Void']>
}
```
I have the following hook wrapper:
```
import { useMutation, useQuery } from '@/gqty'
import {type Mutation } from '@/gqty/schema.generated';
import { UseMutationState } from '@gqty/react'
type KeysMatching<T extends object, V> = {
  [K in keyof T]-?: T[K] extends V ? K : never
}[keyof T];

function useMutation2<M extends KeysMatching<Mutation, object>> (m: M,  cb: (result: ReturnType<Mutation[M]>) => void) {
  
  const [func, state] = useMutation((mutation, inputs) => {
    const mf = mutation[m]
    if (mf != null && typeof mf === 'function') {
      cb(mf(inputs))
    }
  })
  // need to force cast it to unknown, im sure there are better ways
  return [func, state ] as unknown as [(args: Parameters<Mutation[M]>[0]) => void, UseMutationState] 
}
```
and here's how to use it:

https://github.com/gqty-dev/gqty/assets/1062920/4943298a-f2ad-4180-affb-af385218fed1

**Reply by vicary** [reply]
Not sure if I follow.

Do you want to default to the direct import method with your wrapper?

**Reply by hyusetiawan** [reply]
no, I want to type the function call from the returning hook, the `func` from `[func, {isLoading}]` is fully typed with args, whereas using the default `useMutation` it's not, only the mutation object inside the callback is typed. Hope I am making more sense :D

**Reply by vicary** [reply]
This may break some of the backward compatibility, because in V2 we allow the `func` to provide a completely different selection function. But I can see the clarity and ease of use with only one way to use it.

With our V3 already approaching beta, I think it may be inconvenient for others to take this breaking change. But we could definitely create a new hook, e.g. `useSimpleMutation` so you get the type `fn(foo: string)` from `useSimpleMutaiton((mutation, foo: string) => {...})`.

If it turns out great and everyone is using it, we can replace the original one in the next major.
